UICatalog/Scenarios/SyntaxHighlighting.cs

```diff
@@ -63,148 +245,155 @@ private void Quit ()
+		/// <returns>Returns a new instance of the object read from the Json file.</returns>
+		public static T ReadFromJsonFile<T> (string filePath) where T : new()
+		{
+			TextReader reader = null;
+			try {
+				reader = new StreamReader (filePath);
+				var fileContents = reader.ReadToEnd ();
+				return (T)JsonSerializer.Deserialize (fileContents, typeof (T));
+			} finally {
+				if (reader != null) {
+					reader.Close ();
+				}
 			}
+		}
+	}
 
-			private bool IsKeyword (List<Rune> line, int idx)
-			{
-				var word = IdxToWord (line, idx);
+	public static class EventExtensions {
+		public static void ClearEventHandlers (this object obj, string eventName)
+		{
+			if (obj == null) {
+				return;
+			}
+
+			var objType = obj.GetType ();
+			var eventInfo = objType.GetEvent (eventName);
+			if (eventInfo == null) {
+				return;
+			}
+
+			var isEventProperty = false;
+			var type = objType;
+			FieldInfo eventFieldInfo = null;
+			while (type != null) {
+				/* Find events defined as field */
+				eventFieldInfo = type.GetField (eventName, BindingFlags.Static | BindingFlags.Instance | BindingFlags.Public | BindingFlags.NonPublic);
+				if (eventFieldInfo != null && (eventFieldInfo.FieldType == typeof (MulticastDelegate) || eventFieldInfo.FieldType.IsSubclassOf (typeof (MulticastDelegate)))) {
+					break;
+				}
 
-				if (string.IsNullOrWhiteSpace (word)) {
-					return false;
+				/* Find events defined as property { add; remove; } */
+				eventFieldInfo = type.GetField ("EVENT_" + eventName.ToUpper (), BindingFlags.Static | BindingFlags.Instance | BindingFlags.NonPublic);
+				if (eventFieldInfo != null) {
+					isEventProperty = true;
+					break;
 				}
 
-				return keywords.Contains (word, StringComparer.CurrentCultureIgnoreCase);
+				type = type.BaseType;
 			}
 
-			private string IdxToWord (List<Rune> line, int idx)
-			{
-				var words = Regex.Split (
-					new string (line.Select (r => (char)r.Value).ToArray ()),
-					"\\b");
+			if (eventFieldInfo == null) {
+				return;
+			}
 
-				int count = 0;
-				string current = null;
+			if (isEventProperty) {
+				// Default Events Collection Type
+				RemoveHandler<EventHandlerList> (obj, eventFieldInfo);
+				return;
+			}
 
-				foreach (var word in words) {
-					current = word;
-					count += word.Length;
-					if (count > idx) {
-						break;
-					}
-				}
+			if (!(eventFieldInfo.GetValue (obj) is Delegate eventDelegate)) {
+				return;
+			}
+
+			// Remove Field based event handlers
+			foreach (var d in eventDelegate.GetInvocationList ()) {
+				eventInfo.RemoveEventHandler (obj, d);
+			}
+		}
+
+		private static void RemoveHandler<T> (object obj, FieldInfo eventFieldInfo)
+		{
+			var objType = obj.GetType ();
+			var eventPropertyValue = eventFieldInfo.GetValue (obj);
+
+			if (eventPropertyValue == null) {
+				return;
+			}
+
+			var propertyInfo = objType.GetProperties (BindingFlags.NonPublic | BindingFlags.Instance)
```

Comment:
Ugh. This feels very hacky. Can you explain why using reflection here is needed?

Reply:
The reason for using it is because the `SyntaxHighlighting` has two samples (`ApplySyntaxHighlighting` and `ApplyLoadRuneCells`) which each one use different events and we don't know the sequence the user will use them and thus ensure that all events of the previous running is always disposed.